File: app/src/main/java/com/yashwant/newsmodule/ui/news/NewsActivity.java
package com.yashwant.newsmodule.ui.news;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;


import com.mvpdaggerrxjvasample.R;
import com.yashwant.newsmodule.app.NewsApplication;
import com.yashwant.newsmodule.model.Hit;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class NewsActivity extends AppCompatActivity implements NewsView, View.OnClickListener {

    @BindView(R.id.progressBar)
    ProgressBar progressBar;
    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @Inject
    NewsPresenter presenter;
    @BindView(R.id.editSearch)
    EditText editSearch;
    @BindView(R.id.buttonSearch)
    Button buttonSearch;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_news);
        ((NewsApplication) getApplication()).getAppComponent().inject(this);
        ButterKnife.bind(this);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.addItemDecoration(new DividerItemDecoration(this,LinearLayoutManager.VERTICAL));
        presenter.setView(this);

        buttonSearch.setOnClickListener(this);
        progressBar.setVisibility(View.GONE);

    }

    @Override
    public void showLoading() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showNews(List<Hit> newsList) {
        recyclerView.setAdapter(new NewsAdapter(newsList));
        recyclerView.getAdapter().notifyDataSetChanged();

    }

    @Override
    public void showErrorMessage() {
        Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void launchNewsDetail(Hit news) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.parse(news.getUrl()));
            startActivity(intent);
        } catch (ActivityNotFoundException e) {
            e.printStackTrace();
        }


    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public void onClick(View view) {
        presenter.getNews(editSearch.getText().toString());
    }

    class NewsAdapter extends RecyclerView.Adapter<NewsViewHolder> {

        private List<Hit> newsList;
        private Date createdAt;

        NewsAdapter(List<Hit> news) {
            this.newsList = news;
        }

        @Override
        public NewsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(NewsActivity.this);
            return new NewsViewHolder(inflater.inflate(R.layout.news_item_layout, parent, false));
        }

        @Override
        public void onBindViewHolder(NewsViewHolder holder, int position) {
            final Hit newsItem = newsList.get(position);
            holder.getTvNewsName().setText(newsItem.getTitle());
            holder.getTvAuthor().setText("By " + newsItem.getHighlightResult().getAuthor().getValue());
            try {
                SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
                createdAt=formatter.parse(newsItem.getCreatedAt());
            } catch (ParseException e) {
                e.printStackTrace();
            }
            holder.getTvDateCreated().setText("on " + DateUtils.getRelativeTimeSpanString(createdAt.getTime(), new Date().getTime(), 0L, DateUtils.FORMAT_ABBREV_ALL).toString());
            holder.getContainer().setOnClickListener(view -> launchNewsDetail(newsItem));
        }

        @Override
        public int getItemCount() {
            return newsList.size();
        }
    }


}

File: app/src/main/java/com/yashwant/newsmodule/ui/news/NewsActivity.kt
package com.yashwant.newsmodule.ui.news

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.Toast


import com.mvpdaggerrxjvasample.R
import com.yashwant.newsmodule.app.NewsApplication
import com.yashwant.newsmodule.model.Hit

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife

class NewsActivity : AppCompatActivity(), NewsView, View.OnClickListener {

    @BindView(R.id.progressBar)
    internal var progressBar: ProgressBar? = null
    @BindView(R.id.recyclerView)
    internal var recyclerView: RecyclerView? = null
    @Inject
    internal var presenter: NewsPresenter? = null
    @BindView(R.id.editSearch)
    internal var editSearch: EditText? = null
    @BindView(R.id.buttonSearch)
    internal var buttonSearch: Button? = null

    override val context: Context
        get() = this

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_news)
        (application as NewsApplication).appComponent!!.inject(this)
        ButterKnife.bind(this)

        recyclerView!!.layoutManager = LinearLayoutManager(this)
        recyclerView!!.addItemDecoration(DividerItemDecoration(this, LinearLayoutManager.VERTICAL))
        presenter!!.setView(this)

        buttonSearch!!.setOnClickListener(this)
        progressBar!!.visibility = View.GONE

    }

    override fun showLoading() {
        progressBar!!.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        progressBar!!.visibility = View.GONE
    }

    override fun showNews(newsList: List<Hit>) {
        recyclerView!!.adapter = NewsAdapter(newsList)
        recyclerView!!.adapter.notifyDataSetChanged()

    }

    override fun showErrorMessage() {
        Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show()
    }

    override fun launchNewsDetail(news: Hit) {
        try {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(news.url)
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            e.printStackTrace()
        }


    }

    override fun onClick(view: View) {
        presenter!!.getNews(editSearch!!.text.toString())
    }

    internal inner class NewsAdapter(private val newsList: List<Hit>) : RecyclerView.Adapter<NewsViewHolder>() {
        private var createdAt: Date? = null

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
            val inflater = LayoutInflater.from(this@NewsActivity)
            return NewsViewHolder(inflater.inflate(R.layout.news_item_layout, parent, false))
        }

        override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
            val newsItem = newsList[position]
            holder.tvNewsName.text = newsItem.title
            holder.tvAuthor.text = "By " + newsItem.highlightResult!!.author!!.value!!
            try {
                val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
                createdAt = formatter.parse(newsItem.createdAt)
            } catch (e: ParseException) {
                e.printStackTrace()
            }

            holder.tvDateCreated.text = "on " + DateUtils.getRelativeTimeSpanString(createdAt!!.time, Date().time, 0L, DateUtils.FORMAT_ABBREV_ALL).toString()
            holder.container.setOnClickListener { view -> launchNewsDetail(newsItem) }
        }

        override fun getItemCount(): Int {
            return newsList.size
        }
    }


}
